lib/tasks/extraneous_dependencies.rake

```diff
@@ -0,0 +1,101 @@
+require "tasks/helpers/compact_index_tasks_helper"
+
+namespace :extraneous_dependencies do
+  def fetch_spec_deps(full_name)
+    spec_uri = URI("https://rubygems.org/quick/Marshal.4.8/#{full_name}.gemspec.rz")
+    http = Net::HTTP.new(spec_uri.host, spec_uri.port)
+    http.use_ssl = true
+
+    request = Net::HTTP::Get.new(spec_uri.request_uri)
+    res = http.request(request)
+
+    raise StandardError, "fetch deps request for #{full_name} failed #{res.inspect}" unless res.code == "200"
+
+    spec_obj = Marshal.load(Gem::Util.inflate(res.body))
+
+    spec_run_deps = spec_obj.dependencies.map do |s|
+      s.name.downcase if s.type == :runtime && Rubygem.where(name: s.name).present?
+    end.compact.sort
+
+    spec_dev_deps = spec_obj.dependencies.map do |s|
+      s.name.downcase if s.type == :development && Rubygem.where(name: s.name).present?
+    end.compact.sort
```

Comment:
After adding the extra logging I mentioned in other comments, I observed the following results:

```
[extraneous_dependencies:clean] 7/3548 errors
[extraneous_dependencies:clean] 3512/3548 version mismatches (run_deps: 3505, dev_deps: 30)
[extraneous_dependencies:clean] 5147 dependencies deleted
```

I went ahead and investigated the errors. Apparently some dependencies from the marshalled index have  symbolized names, and that was causing a bad symbol vs string comparison. For example, `rumai-3.0.0`:

```
> spec_uri = URI("https://rubygems.org/quick/Marshal.4.8/rumai-3.0.0.gemspec.rz")
=> #<URI::HTTPS https://rubygems.org/quick/Marshal.4.8/rumai-3.0.0.gemspec.rz>

> require 'net/http'
=> true

> http = Net::HTTP.new(spec_uri.host, spec_uri.port)
=> #<Net::HTTP rubygems.org:443 open=false>

> http.use_ssl = true
=> true

> request = Net::HTTP::Get.new(spec_uri.request_uri)
=> #<Net::HTTP::Get GET>

> res = http.request(request)
=> #<Net::HTTPOK 200 OK readbody=true>

> Marshal.load(Gem::Util.inflate(res.body)).dependencies
=> [<Gem::Dependency type=:development name=:dfect requirements=">= 0">, <Gem::Dependency type=:development name="inochi" requirements="~> 1">, <Gem::Dependency type=:runtime name="inochi" requirements="~> 1">]
```

After adding some string conversions to the dependency fetching, the task run with 0 errors:

```diff
diff --git a/lib/tasks/extraneous_dependencies.rake b/lib/tasks/extraneous_dependencies.rake
index 98b19041..53052180 100644
--- a/lib/tasks/extraneous_dependencies.rake
+++ b/lib/tasks/extraneous_dependencies.rake
@@ -14,11 +14,11 @@ namespace :extraneous_dependencies do
     spec_obj = Marshal.load(Gem::Util.inflate(res.body))
 
     spec_run_deps = spec_obj.dependencies.map do |s|
-      s.name.downcase if s.type == :runtime && Rubygem.where(name: s.name).present?
+      s.name.to_s.downcase if s.type == :runtime && Rubygem.where(name: s.name.to_s).present?
     end.compact.sort
 
     spec_dev_deps = spec_obj.dependencies.map do |s|
-      s.name.downcase if s.type == :development && Rubygem.where(name: s.name).present?
+      s.name.to_s.downcase if s.type == :development && Rubygem.where(name: s.name.to_s).present?
     end.compact.sort
 
     [spec_run_deps, spec_dev_deps]
```

Reply:
kinda left symbol issue as is on purpose. `to_s` will avoid raising the error, yes.. however, the versions in question will remain uninstallable. 

do you think using `to_s` is better?

Reply:
the primary purpose of rescue in this rake task is dealing with any network issue.

Reply:
Given the improvements in the number of queries to rubygems.org that deleting the these dependencies seems to have, I was thinking that by deleting the bad dependencies also in this case, we can avoid even more queries. `rumai-3.0.0` doesn't become installable because of this, but if `bundler` resolves `rumai` to a higher version that is installable, deleting these dependencies could speed up `bundle install` in such a situation?

Reply:
Also, by having 0 errors and thus expecting 0 errors when the task is run in production, it becomes simpler to know whether we need to rerun the task I guess.

Reply:
added `to_s`. thanks for explaining.